### DISCUSSION TITLE
Support more TypeScript/JavaScript file extensions

### DISCUSSION THREAD

**Discussion by shrinktofit** [cocos/cocos-engine#11238] [Ideas]
Now that [TypeScript 4.7 was released](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/), it eventually implemented full Node.js ESM support. It introduces new file extensions: _.mts_, _.cts_ which corresponding to _.mjs_, _.cjs_ respectively. We're planing to land TypeScript 4.7 into ~3.6 and It'd be better if we can support these extensions.

- Basic support, including:

  - [ ] Editor recognizes them as script assets(i.e allocates UUIDs to them).

  - [ ] Module resolution algorithm.

    > Note: We have announced that all _*.ts_ files are treated as ES modules. This is also true for _*.mts_ not true for _*.cts_.

- [ ] _.mjs_/_.cjs_ should not be imported as "plugin script".

- [ ] Document and teach.

- [ ] Possible asset creation menu change? Shall we add new menu items into asset creation menu to create _.mts_ or _.cts_ files?

    > Note: we have been preventing developers from creating JavaScript files(_.js_, _.mjs_, _.cjs_) in editor, intentionally.

**Comment by jareguo** [comment]
It doesn't seem to see too obvious a benefit, and we still recommend staying unified at this stage.

For third party modules, we can support importing mts if needed. But it does not feel necessary to create mts scripts in editor.
